pkg/cloudsources/ocm/client.go

```diff
@@ -0,0 +1,127 @@
+package ocm
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	gogoProto "github.com/gogo/protobuf/types"
+	sdkClient "github.com/openshift-online/ocm-sdk-go"
+	accountsmgmtv1 "github.com/openshift-online/ocm-sdk-go/accountsmgmt/v1"
+	pkgErrors "github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/clientconn"
+	"github.com/stackrox/rox/pkg/cloudsources/discoveredclusters"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/urlfmt"
+)
+
+type ocmClient struct {
+	conn          *sdkClient.Connection
+	cloudSourceID string
+}
+
+// NewClient creates a client to interact with OCM APIs.
+func NewClient(config *storage.CloudSource) (*ocmClient, error) {
+	connection, err := sdkClient.NewConnectionBuilder().
+		URL(urlfmt.FormatURL(config.GetOcm().GetEndpoint(), urlfmt.HTTPS, urlfmt.NoTrailingSlash)).
+		Tokens(config.GetCredentials().GetSecret()).Agent(clientconn.GetUserAgent()).Build()
+
+	if err != nil {
+		return nil, pkgErrors.Wrap(err, "creating OCM connection")
+	}
+
+	return &ocmClient{
+		conn:          connection,
+		cloudSourceID: config.GetId(),
+	}, nil
+}
+
+func (c *ocmClient) GetDiscoveredClusters(ctx context.Context) ([]*discoveredclusters.DiscoveredCluster, error) {
+	var (
+		subs  []*accountsmgmtv1.Subscription
+		total int
+		page  = 1
+	)
+
+	// Taken from console.redhat.com/openshift.
+	// Filter for all subscriptions which have:
+	//	- a cluster associated with it,
+	// 	- the cluster is a valid OpenShift plan
+	//	- the status is "Active"
+	//  - and name / cluster_id / external_cluster_id are given.
+	subscriptionSearch := "(cluster_id!='') " +
+		"AND (plan.id NOT IN ('RHACS', 'RHACSTrial', 'RHOSR', 'RHOSRTrial', 'RHOSAK', 'RHOSAKTrial', 'RHOSE', 'RHOSETrial')) " +
```

Comment:
We'll be vulnerable to the same bug we had with acs-cs where Centrals suddenly appeared as OpenShift clusters in the OCM list. Maybe a whitelist is the safer choice here in case new products get added?

Reply:
While this now has the disadvantage that we can miss potential new plans (e.g. a new flavor of hosted control plane), I've convinced myself that it's easier to say "ACS will support plan X for OCM cloud sources to be discovered" compared to "ACS has yet again a bug where it shows some plans I cannot secure with ACS (e.g. Ansible Wisdom)".

The list of plans has been taken from [here](https://telemeter-lts-dashboards.datahub.redhat.com/explore?orgId=1&left=%7B%22datasource%22:%22P034A7E52C5C9534B%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P034A7E52C5C9534B%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22count%20by%20%28product%29%20%28ocm_subscription%29%22,%22legendFormat%22:%22__auto%22,%22range%22:false,%22instant%22:true,%22exemplar%22:false%7D%5D,%22range%22:%7B%22from%22:%22now-5m%22,%22to%22:%22now%22%7D%7D) which seems to be the complete list.